part-1.py

```diff
@@ -4,17 +4,30 @@
 # the appropriate comment.
 
 # factorial
-
-
+def factorial(n):
+    if n < 0:
+        raise ValueError
+    if n == 0:
+        return 1
+    return n * factorial(n-1)
 
 # reverse
-
-
+def reverse(text):
+    if len(text) == 0:
+        return text
+    return text[-1] + reverse(text[:len(text)-1])
```

Comment:
Minor note: using `-1` could make this more readable.
```suggestion
    return text[-1] + reverse(text[:-1])
```